### PR TITLE
feat(semantic): align class scope with typescript

### PR DESCRIPTION
**Pull request by Dunqing**
```ts
class Klass <T>   extends Root       <R>                   {}
//    ^^^^^ ^^^           ^^^^       ^^^                   ^^
//    id type_paramter super_class super_type_parameters  body
```
I reorder fields according to the order above

The class scope is not defined in the spec. But we need to create a scope for `class` to store `TypeParamters`

### CODE REVIEW

File: tasks/coverage/parser_test262.snap [hunk 1]
```diff
@@ -2,12 +2,186 @@ commit: a1587416
...
 Expect Syntax Error: "language/import/import-assertions/json-invalid.js"
 Expect Syntax Error: "language/import/import-assertions/json-named-bindings.js"
 Expect Syntax Error: "language/import/import-attributes/json-invalid.js"
 Expect Syntax Error: "language/import/import-attributes/json-named-bindings.js"
+Expect to Parse: "built-ins/Function/prototype/toString/class-expression-explicit-ctor.js"
+
+  × Identifier `A` has already been declared
+    ╭─[built-ins/Function/prototype/toString/class-expression-explicit-ctor.js:10:5]
+  9 │ 
+ 10 │ let A = /* before */class /* a */ A /* b */ extends /* c */ B /* d */ { /* e */ constructor /* f */ ( /* g */ ) /* h */ { /* i */ ; /* j */ } /* k */ m /* l */ ( /* m */ ) /* n */ { /* o */ } /* p */ }/* after */;
```
Comment (Boshen) [inline tasks/coverage/parser_test262.snap:16]: Oh no!

Comment (Dunqing) [reply tasks/coverage/parser_test262.snap:16]: Don't worry! It has already been fixed in https://github.com/oxc-project/oxc/pull/4141

Comment (Boshen) [reply tasks/coverage/parser_test262.snap:16]: Can we merge these 2 PRs, this is confusing :-/

**Comment by overlookmotel** [comment]
> The class scope is not defined in the spec

I think it is - please see #4161.

File: tasks/coverage/parser_test262.snap (hunk 1 above)
Comment (Dunqing) [reply tasks/coverage/parser_test262.snap]: Done

**Comment by Dunqing** [comment]
Blocked on https://github.com/oxc-project/oxc/issues/4200

**Comment by Dunqing** [comment]
The regression tests have been fixed in #4156. Should I merge into this PR?

**Review by Dunqing — COMMENTED** [review]
Not adding symbols to scope may cause some confusion. For example, in mangler, we need to rename bindings, but we can't find the symbols for function expression and class expression in scope.

 I'll revert this change

File: crates/oxc_semantic/src/builder.rs [hunk 2]
```diff
@@ -1586,6 +1546,20 @@ impl<'a> SemanticBuilder<'a> {
...
+                if let Some(node) = self.nodes.parent_node(self.current_node_id) {
+                    if let AstKind::Class(class) = node.kind() {
+                        if class.is_expression() {
+                            // We must to bind class expression when enter BindingIdentifier,
+                            // because we should add binding to current scope
+                            let previous_node_id = self.current_node_id;
+                            self.current_node_id = node.id();
+                            class.bind(self);
+                            self.current_node_id = previous_node_id;
+                        }
```
Comment (Dunqing) [inline crates/oxc_semantic/src/builder.rs:1559]: Bind `ClassExpression` in `BindingIdentifier` to add the binding to the correct scope

**Pull request merged**